Add MenuItem.InputGestureText.

grokys opened the pull request:
## What does the pull request do?

Adds `MenuItem.InputGestureText` as implemented in WPF. This can be used to display an input gesture alongside a menu item. Note that it does _not_ implement handling for that gesture. We need to work out what we're going to do there as [`HotKey` is broken for `MenuItem`](https://github.com/AvaloniaUI/Avalonia/issues/2441).

![image](https://user-images.githubusercontent.com/1775141/75089611-df4cee80-555a-11ea-81e0-64aa3ab825a3.png)


## How was the solution implemented (if it's not obvious)?

There's a nasty hack in here that I had to port from WPF, due to WPF's `Grid` (which is now our `Grid`) being unfit for purpose in this case:

https://github.com/AvaloniaUI/Avalonia/compare/feature/menuitem-inputgesturetext?expand=1#diff-d6410592bff9becfeb79af46877af0b2R336-R346

We probably should fix `Grid` to work properly in this case, but that's another PR.

grokys committed:
Don't add extra pixel to AccessText measurement.

It's not needed; the underscore can be drawn in the descender space.

grokys committed:
Add MenuItem.InputGestureText.

grokys committed:
Show input gesture text in control catalog.

jmacato committed:
Merge branch 'master' into feature/menuitem-inputgesturetext

kekekeks commented:
We also need to update [NativeMenuBar template](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Default/NativeMenuBar.xaml)

kekekeks commented on src/Avalonia.Controls/MenuItem.cs:
```diff
@@ -306,6 +327,36 @@ protected override void OnDetachedFromLogicalTree(LogicalTreeAttachmentEventArgs
+            {
+                // HACK: This nasty but it's all WPF's fault. Grid uses an inherited attached
+                // property to store SharedSizeGroup state, except property inheritance is done
+                // down the logical tree. In this case, the control which is setting
```
We could add inheritance policy option in property registration

grokys replied:
I think that would be complicating things too much. An inherited property here is simply the wrong way to do it IMO.

MarchingCube commented:
Just to add my five cent about https://github.com/AvaloniaUI/Avalonia/issues/2441 which was reported by my colleague - having actual gesture handling on `MenuItem` makes no sense because:
- gesture actually is not supposed to trigger when you have menu open
- when you have menu closed - how it is supposed to even handle gestures if it is not on the event bubbling path?
- what if you build your menu dynamically and there are no `MenuItems` apart from the root ones.

Our approach was to implement our own "Routed command" like architecture so menu items can just request command execution but actual execution is handled somewhere higher in the tree. This also allows us to implement multiple scopes for commands so raising gestures can be context dependent.

MarchingCube commented:
Just to add my five cent about https://github.com/AvaloniaUI/Avalonia/issues/2441 which was reported by my colleague - having actual gesture handling on `MenuItem` makes no sense because:
- gesture actually is not supposed to trigger when you have menu open
- when you have menu closed - how it is supposed to even handle gestures if it is not on the event bubbling path?
- what if you build your menu dynamically and there are no `MenuItems` apart from the root ones.

Our approach was to implement our own "Routed command" like architecture so menu items can just request command execution but actual execution is handled somewhere higher in the tree. This also allows us to implement multiple scopes for commands so raising gestures can be context dependent.

MarchingCube commented:
Just to add my five cent about https://github.com/AvaloniaUI/Avalonia/issues/2441 which was reported by my colleague - having actual gesture handling on `MenuItem` makes no sense because:
- gesture actually is not supposed to trigger when you have menu open
- when you have menu closed - how it is supposed to even handle gestures if it is not on the event bubbling path?
- what if you build your menu dynamically and there are no `MenuItems` apart from the root ones.

Our approach was to implement our own "Routed command" like architecture so menu items can just request command execution but actual execution is handled somewhere higher in the tree. This also allows us to implement multiple scopes for commands so raising gestures can be context dependent.

MarchingCube commented:
Just to add my five cent about https://github.com/AvaloniaUI/Avalonia/issues/2441 which was reported by my colleague - having actual gesture handling on `MenuItem` makes no sense because:
- gesture actually is not supposed to trigger when you have menu open
- when you have menu closed - how it is supposed to even handle gestures if it is not on the event bubbling path?
- what if you build your menu dynamically and there are no `MenuItems` apart from the root ones.

Our approach was to implement our own "Routed command" like architecture so menu items can just request command execution but actual execution is handled somewhere higher in the tree. This also allows us to implement multiple scopes for commands so raising gestures can be context dependent.

MonkAlex commented:
Its WPF realization?
In crossplatform i want use ⌘-O (macos) and Ctrl+O (windows), how to write it in InputGestureText?

MarchingCube commented on src/Avalonia.Controls/MenuItem.cs:
```diff
@@ -194,6 +202,19 @@ public object Icon
+        /// Setting this property does not cause the input gesture to be handled by the menu item,
+        /// it simply displays the gesture text in the menu.
+        /// </remarks>
+        public object InputGestureText
```
I guess it was supposed to be `string` and not `object` ?

grokys replied:
D'oh! Not sure how that happened.

grokys replied:
Fixed

MarchingCube reviewed (approved):
LGTM with one minor issue.

MarchingCube committed:
Merge branch 'master' into feature/menuitem-inputgesturetext

grokys commented:
@MonkAlex good question. As far as I know there's no way to automatically translate key bindings between platforms, so it would have to be done manually and I don't think we would want to handle this at the level of individual controls.

Xamarin Forms for example has [`OnPlatform`](https://xamarinhelp.com/xamarin-forms-onplatform-runtimeplatform/) for this.

In the meantime I think it'd be best to suggest that, if you need platform-specific values for this, to set up the key bindings in code and bind to those.

grokys committed:
Use correct property type.

grokys committed:
Display gestures in NativeMenuBar.

To do this we needed to change the `KeyGesture.ToString()` method to produce input gesture text suitable for menus. Also demonstrated in `MainWindow` how to produce different gestures/headers depending on platform.

grokys commented:
@kekekeks updated to display gestures in `NativeMenuBar`. To do this I needed to change the `KeyGesture.ToString()` method to produce input gesture text suitable for menus. Also demonstrated in `MainWindow` how to produce different gestures/headers depending on platform.

Regarding the "Quit" menu being different between OSX and Windows, our `PlatformHotkeyConfiguration` probably needs to be improved to allow specifying that it's "Alt+F4" on Windows and "Cmd+Q" on OSX, but that's another PR I think.

cc: @danwalmsley to do a sanity check on OSX.

MarchingCube commented on src/Avalonia.Input/KeyGesture.cs:
```diff
@@ -108,19 +109,43 @@ public static KeyGesture Parse(string gesture)
+                if (s.Length > 0)
                 {
-                    parts.Add(flag.ToString());
+                    s.Append("+");
```
`char` variant is faster :)

MarchingCube commented on src/Avalonia.Input/KeyGesture.cs:
```diff
@@ -108,19 +109,43 @@ public static KeyGesture Parse(string gesture)
+            if (KeyModifiers.HasFlagCustom(KeyModifiers.Meta))
+            {
+                Plus(s);
+                s.Append("⌘");
```
I guess this is platform dependent, on Windows this should be windows key.

grokys replied:
Hmm, you're right. I don't really want to add platform-specific code to a `ToString` method, that feels like asking for trouble. Should we really be mapping the Windows key to Meta? Is there precedent for this? Feels like this should be a different key modifier to me.

kekekeks replied:
There is no "windows" key in Qt and GTK. Just Meta/Super

grokys replied:
Ok, so we want to follow Qt/GTK here is what you're saying (rather than make it a separate key in Avalonia)?

If so, will we need to format this text differently depending on platform, meaning simply calling `ToString()` isn't a good idea. Which means that maybe having `InputGestureText` as a string isn't a good idea.

kekekeks commented:
@grokys 
>As far as I know there's no way to automatically translate key bindings between platforms, so it would have to be done manually and I don't think we would want to handle this at the level of individual controls.

We currently have some mappings in https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Input/Platform/PlatformHotkeyConfiguration.cs#L12

The idea was to parse `Cmd` in a platform-specific way, we've discussed that with @danwalmsley , but it wasn't implemented for some reason.

grokys commented:
@kekekeks yeah, I think that can be done automatically to an extent, but not in all cases, for example the `Exit`/`Quit`/`Cmd+Q`/`F4` situation that I already outlined.

I think we can solve this problem, but it should be separate from this PR.

grokys committed:
Merge branch 'master' into feature/menuitem-inputgesturetext

grokys committed:
MenuItem.InputGestureText -> InputGesture

Changed `MenuItem`'s `string InputGestureText` property to `KeyGesture InputGesture` and gave it a value converter in the default template, because the string representation of a key gesture depends on platform, and putting platform-specific behavior into `ToString` isn't a good idea.

grokys commented:
Ok, because of the discussion in https://github.com/AvaloniaUI/Avalonia/pull/3602#discussion_r385020016 I've changed tack slightly:

- Changed `InputGestureText` into `InputGesture` and made it take a `KeyGesture` instead of a `string`
- Added a `PlatformKeyGestureConverter` that outputs the string representation of a `KeyGesture` depending on platform
- Use the converter in the `MenuItem` template

Does this sound like a better way to do things? I didn't want to make the result of `KeyGesture.ToString()` change depending on what platform you're running on.

MarchingCube commented:
![image](https://user-images.githubusercontent.com/2588062/75933437-4d0dea00-5e7a-11ea-80cc-6d94adeae71f.png)

Submenus behave in a weird way - it seems that your hack for grid scopes might not be working as expected :)

And after thinking about that platform specific gesture conversions. It is a good idea but I guess we need to handle keys in a bit nicer way since even in my example above you can see `Ctrl+OemTilde` and most of the software would show this a `Ctrl+~` or in whatever human readable format (so `/` and not `OemPipe`). From my quick glance in OSX some keys even use emojis in some example menus (like backspace). Saying that partially from the experience since our project had to roll simple human readable key gesture conversion to not confuse people with raw enum names.

grokys commented:
Interesting, I was about to say "submenus work fine here" but then I discovered that I can reproduce the problem you screenshotted above by dragging the window between screens with a different DPI. Will investigate.

Good point about `OenTilde` etc, we need to convert those to the proper symbol. Regarding OSX, is there a guide for how to display shortcuts? I've done a search but been unable to fine one.

grokys committed:
Improve platform display of KeyGesture.

grokys committed:
Detach grid hack on logical tree detachment.

Doing it on visual tree detachment caused the menu to bug out when showing a submenu more than once.

grokys commented:
Ok, I think I've addressed those issues @MarchingCube.

jmacato committed:
Merge branch 'master' into feature/menuitem-inputgesturetext

MarchingCube committed:
Merge branch 'master' into feature/menuitem-inputgesturetext

MarchingCube commented on src/Avalonia.Themes.Default/MenuItem.xaml:
```diff
@@ -36,12 +47,15 @@
                 </DataTemplate>
               </ContentPresenter.DataTemplates>
             </ContentPresenter>
+            <TextBlock Grid.Column="3"
```
I think having a name assigned here will make styling easier. In our case we would like this text to have different color.

grokys replied:
Done!

MarchingCube commented on src/Avalonia.Controls/Converters/PlatformKeyGestureConverter.cs:
```diff
@@ -0,0 +1,173 @@
+
+namespace Avalonia.Controls.Converters
+{
+    public class PlatformKeyGestureConverter : IValueConverter
```
Wondering if that could be exposed as a static helper since I would like to reuse that logic in our app and I would need to create a new instance of this converter to do so otherwise (I can obviously use a static shared instance but that feels hacky).

grokys replied:
Done!

grokys committed:
Merge branch 'master' into feature/menuitem-inputgesturetext

grokys committed:
Give input gesture TextBlock a name.

grokys committed:
Expose platform KeyGesture conversion as static method.

grokys committed:
Added a few more keys to key gesture converter.

grokys committed:
Allow overriding inner menu padding.

MarchingCube committed:
Merge branch 'master' into feature/menuitem-inputgesturetext

grokys committed:
Use words for arrow keys.

Gillibald commented on src/Avalonia.Controls/Converters/PlatformKeyGestureConverter.cs:
```diff
@@ -148,6 +148,8 @@ private static string ToString(Key key)
                 Key.D9 => "9",
                 Key.Decimal => ".",
                 Key.Divide => "/",
+                Key.Down => "Down Arrow",
```
Why are we using text here instead of symbols?

grokys replied:
Because as far as we could tell, on Windows at last, menus use e.g. `Ctrl+Up Arrow` as the key gesture text.
